### PR TITLE
Use XDG config paths on FreeBSD

### PR DESCRIPTION
**Pull request by emaste**
### Description
We should use XDG config paths on FreeBSD. This could be done by adding `USE_XDG` in the os-freebsd.cmake files, but this would leave the `#else` case in `#ifdef USE_XDG` unused. It is also currently broken. Thus, just remove the `#else` case and `USE_XDG` references.

Also reintroduce `move_to_xdg` from ba02e065fe91 to migrate FreeBSD installs to XDG paths on upgrade.

### Motivation and Context
OBS currently fails to start if built without `USE_XDG`.
We should be following XDG config paths on FreeBSD anyway.

### How Has This Been Tested?
OBS starts correctly on FreeBSD with these patches applied. I ran `rm -rf ~/.obs-studio/ ~/.config/obs-studio/`, started again, and confirmed that default config is created under ~/.config/obs-studio. Removed the directories again, created config under ~/.obs-studio, started again, and confirmed that the config was moved to ~/.config/obs-studio.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by emaste** [comment]
There is some support for OpenBSD and DragonflyBSD in OBS, but cmake os-specific support exists only for Windows, macOS, Linux, and FreeBSD. The non-XDG case ($HOME/.obs-studio/ config path) doesn't work today anyway, and I suspect we want to use the XDG config paths on all non-Windows/macOS OSes.

I'll try to find OpenBSD and DragonFlyBSD folks to comment here / collaborate.

**Comment by emaste** [comment]
I need to update the commit message on the first commit, as I wrote it before reintroducing `move_to_xdg` for FreeBSD (only). With both commits in this pull request config will be moved automatically.

Also CC @norihiro, who removed `move_to_xdg` in c0532168243b78a7f880b3ef0c1b20cfe9be51ca.

**Comment by RytoEX** [comment]
I'm not sure that this will land in the 31 beta unless it's demonstrated that this is a regression there.

**Comment by emaste** [comment]
A new installation on FreeBSD from git master will not start, and I believe this is a regression from 2635cf3a2a.

That said, I'm somewhat indifferent to timing on resolving this - almost all FreeBSD OBS users will be installing OBS from our ports collection, and the port maintainer can apply this as a local patch if it does not land upstream in time.

File: libobs/util/platform-nix.c
Comment (RytoEX) [inline libobs/util/platform-nix.c:301]: If removing the XDG code path, should this comment be updated?

Comment (emaste) [reply libobs/util/platform-nix.c]: Yes

**Comment by alarixnia** [comment]
We (NetBSD) have an obs package but never upstreamed the patches. Ideally the XDG paths should be used on every Unix-like system.

**Review by GeorgesStavracas — APPROVED** [review]
I like this, very much actually. Thanks @emaste!

**Review by kkartaltepe — APPROVED** [review]

**Comment by kkartaltepe** [comment]
I feel the migration code should probably just remain a freebsd patch personally, but its fine overall.

**Review by tytan652 — APPROVED** [review]
LGTM

File: UI/obs-app.cpp
Comment (PatTheMav) [inline UI/obs-app.cpp:1514]: I really dislike adding boilerplate code for a single version upgrade that - as experience has taught me - will then linger on as dead code for years, especially as it perpetuates the patchwork of our function bodies (there really should be a single "InitPlatform()` method on `OBSApp` that uses a platform-independent API and uses platform-dependent source files instead of littering our functions with preprocessor conditions all over the place, it's just messy all around).

That said I'm fine with this as a quick fix if a TODO is removed with a version target so that this will be removed after a majority of users has probably been automatically migrated and a commitment to remove it then.

File: UI/obs-app.cpp
Comment (PatTheMav) [inline UI/obs-app.cpp:1536]: Usually we opt to:

* Not delete old files (just in case people want to downgrade or recover from other issues)
* And _if_ we choose to rename we explicitly copy first, then delete (and not do it in a single step)

File: UI/obs-app.cpp
Comment (PatTheMav) [inline UI/obs-app.cpp:2499]: Wouldn't this be optionally conditional on whether portable mode is enabled as a build flag and portable mode _not_ being selected?

Because with portable mode enabled and selected, only config files relative to OBS' location are to be loaded. I'd consider it incorrect behaviour if the app interferes with configuration files that it should not even be aware of/not even consider in that case.

If I understood it right, it's only when _not_ running in portable mode that the new XDG-influenced location would come into play.

File: UI/obs-app.cpp
Comment (emaste) [reply UI/obs-app.cpp:1514]: You're correct, this code was originally introduced in 2015 in ba02e065fe91 and persisted until 2023 (removed in c0532168243b78a7f880b3ef0c1b20cfe9be51ca).

How about keeping it for two major versions? I'll add `/* TODO: Remove after version 32.0. */`

Comment (PatTheMav) [reply UI/obs-app.cpp:1514]: Yeah I kinda thought about keeping it for a full major version cycle as well.

File: UI/obs-app.cpp
Comment (emaste) [reply UI/obs-app.cpp:2499]: It turns out that the FreeBSD cmake bits don't have an option to enable portable mode at the moment - it seems there are a number of changes that need to be brought over from the Linux config (e.g. #11410).

I can follow up with a change to enable portable mode builds on FreeBSD and call this only when portable mode is not selected.

File: UI/obs-app.cpp
Comment (emaste) [reply UI/obs-app.cpp:1536]: I restored the code that previously existed to handle this transition on Linux, and would rather leave it as is. If it's not sufficient now (and I understand if the bar is higher than in 2015) I'd rather just leave it out, and just mention the config path change in a release note.

For the second bullet, if we do keep this migration then a recursive copy-and-delete seems more error-prone than the atomic rename.

File: UI/obs-app.cpp
Comment (PatTheMav) [reply UI/obs-app.cpp:2500]: If there's no way to enable it on FreeBSD than this should indeed be sufficient.

File: UI/obs-app.cpp
Comment (PatTheMav) [reply UI/obs-app.cpp:1537]: As this affects FreeBSD exclusively, I'll leave this to your judgement.

**Comment by PatTheMav** [comment]
> I'll try to find OpenBSD and DragonFlyBSD folks to comment here / collaborate.

As far as CMake is concerned we through all BSD variants into the same bucket. IIRC there are distinctions between BSD variants in some parts of the "Linux"-themed platform code.

Also any mentions of DragonFly were removed from CMake in the last rework.

File: UI/obs-app.cpp
Comment (emaste) [reply UI/obs-app.cpp:1537]: In that case my preference is to proceed with the pull request as is. When I created the "Remove non-XDG code" change I expected that (the few FreeBSD OBS) users would have to just move their config manually (and originally mentioned that in the commit message). A bit later I found the `move_to_xdg` function that was written years ago for exactly the same upgrade case in Linux.

**Review by RytoEX — APPROVED** [review]
Even if we do have to restore the `move_to_xdg` migration function for a major release or two, this seems fine.

**Pull request merged**
